simphony/testing/abc_check_engine.py

```diff
@@ -35,173 +35,191 @@ def create_dataset_items(self):
-        filename = os.path.join(self.temp_dir, 'test.cuds')
-        with closing(H5CUDS.open(filename)) as handle:
-            handle.add_dataset(self.create_dataset(name='test'))
-            ds = handle.get_dataset("test")
+    def add_operation(self, container, *args, **kwrds):
+        method = getattr(
+            sys.modules[__name__],
+            self.operation_mapping['add item'])
+        return method(*args, **kwrds)
 
-            self.assertEqual("test", ds.name)
-            self.check_emtpy_dataset(ds)
+    def test_get_missing_dataset(self):
+        handle = self.engine_factory()
+        with self.assertRaises(ValueError):
+            handle.get_dataset('foo')
 
-    def test_add_dataset_empty_data(self):
-        filename = os.path.join(self.temp_dir, 'test.cuds')
-        with closing(H5CUDS.open(filename)) as handle:
-            handle.add_dataset(self.create_dataset(name='test'))
-            pc = handle.get_dataset("test")
-            self.assertEqual(DataContainer(), pc.data)
-            self.assertEqual(0, len(pc.data))
+    def test_add_dataset(self):
+        handle = self.engine_factory()
+        reference = self.create_dataset(name='test')
+        handle.add_dataset(reference)
+        ds = handle.get_dataset("test")
+
+        self.compare_operation(reference, ds, testcase=self)
 
     def test_add_dataset_data_copy(self):
-        filename = os.path.join(self.temp_dir, 'test.cuds')
-        with closing(H5CUDS.open(filename)) as handle:
-            handle.add_dataset(self.create_dataset(name='test'))
-            pc = handle.get_dataset("test")
-            data = pc.data
-            data[CUBA.NAME] = 'somename'
-
-            # Since the returned data is always a copy,
-            #  therefore the pc.data should not have changed
-            self.assertNotIn(CUBA.NAME, pc.data)
-            # And the length should be still zero
-            self.assertEqual(0, len(pc.data))
-            pc.data = data
-            # This time we replaced the pc.data,
-            #  therefore it should have been changed
-            self.assertIn(CUBA.NAME, pc.data)
-            # The length also should have been changed
-            self.assertEqual(1, len(pc.data))
+        handle = self.engine_factory()
+
+        reference = self.create_dataset(name='test')
+
+        reference_data = reference.data
+        reference_data[CUBA.NAME] = 'foo_name'
+        reference.data = reference_data
+
+        handle.add_dataset(reference)
+
+        ds = handle.get_dataset('test')
+
+        data = ds.data
+        data[CUBA.NAME] = 'somename'
+        data[CUBA.MATERIAL_ID] = 0
+
+        # Since the returned data is always a copy,
+        #  therefore the ds.data should not have changed
+        self.assertNotEqual(data[CUBA.NAME], ds.data[CUBA.NAME])
+        self.assertNotIn(CUBA.MATERIAL_ID, ds.data)
+        # And the length should be still one
+        self.assertEqual(1, len(ds.data))
```

Comment:
I belive we can keep those asserts if we just query the size of the original DataContainer, something like:

```python 
reference = self.create_dataset(name='test')
reference_data = reference.data
reference_datasize = len(reference_data)

...

self.assertEqual(reference_datasize , len(ds.data))

...

self.assertEqual(reference_datasize + 1, len(ds.data))

```

Along with the other tests it should guarantee that the data is the same and no extra data has been added in the process. This is important since, as you mentioned we cannot relay on the data inside the datatest being empty ( it may happens that it already contains CUBA.MATERIAL_ID).

Furthermore, I would also suggest to use our own DataContainer created inside the test, but this conflicts somehow with the purpose of the test.